inc/class-wp-forms-api.php

```diff
@@ -713,6 +737,11 @@ static function render_element( $element, &$values, $form = null ) {
 			$markup .= $element['#label_position'] == 'after' ? $label : '';
 		}
 
+		// Add validation message if element value is not valid
+		if( !$element['#valid'] ) && $element['#invalid_message'] ) {
+			$markup .= self::make_tag( 'span', array( 'class' => 'wp-form-element-invalid-message' ), $element['#invalid_message'] );
```

Comment:
See I think you should be able to have a different message for each validation check so you can give targeted feedback about exactly what is wrong with the value. Also I think there should be an option for the form to either default back to the previously saved value or default (if there is one); OR show the incorrect value that the user tried to submit. Because I think the latter is more friendly for the user, but it currently employs the former.